sandbox/samples/cpp/alembicassembly/alembicassembly.cpp

```diff
@@ -0,0 +1,833 @@
+
+// appleseed.renderer headers.
+//#include "renderer/api/bsdf.h"
+#include "renderer/api/object.h"
+//#include "renderer/api/material.h"
+#include "renderer/api/project.h"
+#include "renderer/api/scene.h"
+#include "renderer/api/utility.h"
+//#include "renderer/api/log.h"
+
+// appleseed.foundation headers.
+#include "foundation/math/matrix.h"
+#include "foundation/math/triangulator.h"
+#include "foundation/math/transform.h"
+#include "foundation/math/vector.h"
+#include "foundation/utility/autoreleaseptr.h"
+#include "foundation/utility/job/iabortswitch.h"
+#include "foundation/utility/string.h"
+
+// appleseed.main headers.
+#include "main/dllvisibility.h"
+
+// AlembicAssembly headers.
+#include "alembicassembly.h"
+
+#include <Alembic/Abc/All.h>
+#include <Alembic/AbcCoreFactory/All.h>
+#include <Alembic/AbcGeom/All.h>
+
+// Standard headers.
+#include <cstddef>
+#include <string>
+#include <vector>
+
+namespace asf = foundation;
+namespace asr = renderer;
+
+namespace
+{
+
+// roll into given matrix stack and return the flatten one
+asf::Matrix4d flatten_xform(std::vector<asf::Matrix4d> mtx_stack)
+{
+    // the matrix we will return
+    auto out_mtx = asf::Matrix4d::identity();
+
+    // mutiply the each matrix of the stack from bottom to top
+    for(auto mtx : mtx_stack)
+    {
+        out_mtx = mtx * out_mtx;
+    }
+
+    return out_mtx;
+}
+
+// roll into given xform sequence stack and return the flatten one
+asr::TransformSequence flatten_xform_seq(const std::vector<asr::TransformSequence>& xform_seq_stack)
+{
+    auto out_xform_seq = asr::TransformSequence();
+
+    if(xform_seq_stack.size() == 1)
+    {
+        // we don't need to flatten anything
+        // TODO: Maybe we can return here?
+        out_xform_seq = xform_seq_stack.back();
+    }
+    else
+    {
+        for(auto xform_seq : xform_seq_stack)
+        {
+            out_xform_seq = xform_seq * out_xform_seq;
+        }
+    }
+    out_xform_seq.optimize();
+    out_xform_seq.prepare();
+
+    return out_xform_seq;
+}
+
+// return if given time samples are linearly time spaced
+bool linearly_sampled(const Alembic::Abc::chrono_t start_time,
+                      const Alembic::Abc::chrono_t end_time,
+                      std::vector<Alembic::Abc::chrono_t> samples)
+{
+    std::cout << "samples.size() " << samples.size() << std::endl;
+    assert(samples.size() > 1);
+
+    // compute time supposed to be between each samples if samples are linearly
+    // sampled.
+    const auto increm = (end_time - start_time) / (samples.size()-1);
+    std::cout << "increm " << increm << std::endl;
+
+    // start sample
+    auto t_accum = samples[0];
+
+    const Alembic::Abc::chrono_t epsilon = 1.0 / 10000.0;
+
+    // get each sample time and compare with a linearly sampled time
+    for(const auto t : samples)
+    {
+        std::cout << t << " " << t-t_accum << " " << (std::fabs(t-t_accum) > epsilon) << std::endl;
+        if (std::fabs(t-t_accum) > epsilon)
```

Comment:
We have more robust ways to compare floating point values:

To check if `x` and `y` are approximately equal:

    foundation::feq(x, y)

Same thing with a custom epsilon value:

    foundation::feq(x, y, 1.0e-6)

To check if `x` is approximately zero:

    foundation::fz(x)